pkg/ha/ha_parser.go

```diff
@@ -0,0 +1,160 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package ha
+
+import (
+	"fmt"
+
+	promModel "github.com/prometheus/common/model"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgmodel/common/errors"
+	"github.com/timescale/promscale/pkg/pgmodel/model"
+	"github.com/timescale/promscale/pkg/prompb"
+)
+
+type haParser struct {
+	service *Service
+}
+
+func NewHAParser(service *Service) *haParser {
+	return &haParser{
+		service: service,
+	}
+}
+
+// ParseData parses timeseries into a set of samplesInfo infos per-metric.
+// returns: map[metric name][]SamplesInfo, total rows to insert
+// When Prometheus & Promscale are running HA mode the below parseData is used
+// to validate leader replica samples & ha_locks in TimescaleDB.
+func (h *haParser) ParseData(tts []prompb.TimeSeries) (map[string][]model.SamplesInfo, int, error) {
+	if len(tts) == 0 {
+		return nil, 0, nil
+	}
+
+	dataSamples := make(map[string][]model.SamplesInfo)
+	rows := 0
+
+	s, _, err := model.LabelProtosToLabels(tts[0].Labels)
+	if err != nil {
+		return nil, rows, err
+	}
+	replicaName := s.GetReplicaName()
+	clusterName := s.GetClusterName()
+
+	if err := checkClusterAndReplicaLabelAreSet(clusterName, replicaName); err != nil {
+		return nil, rows, err
+	}
+
+	// find samples time range
+	var (
+		minTUnix, maxTUnix int64
+		minTWasSet         = false
+	)
+	for i := range tts {
+		t := &tts[i]
```

Comment:
Local vars should get erased on compilation, and everything ends up being a register address. 
On the other hand it may help with not having to de-reference the pointer on each access if the compiler doesn't optimize that also